### PR TITLE
Headline Transformer

### LINKED ISSUE #3930: MAP - Headline Transformer
**Is your feature request related to a problem? Please describe.**
The response we get from ARES needs to be transformed into something we can render.  This ticket is for the headlines

**Describe the solution you'd like**
Collect scenarios of what is possible within the CPS headlines.  Write a function that standardises them to an output that we can render

**Describe alternatives you've considered**
N/A

**Testing notes**
[Tester to complete]

Dev insight: Test automation is sufficient - this ticket will have no user-facing effect

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by ryanmccombe**
Resolves https://github.com/bbc/simorgh/issues/3930

**Overall change:** Convert the CPS headline block into something we can render

**Code changes:**

- Add new transformer with unit tests

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] This PR requires manual testing

### CODE REVIEW

**Review by dr3 — COMMENTED** [review]
Howcome this isnt using a preprocessor rule? We need a generic way to parse xml, rather than doing it case by case in containers

**Comment by ryanmccombe** [comment]
> Howcome this isnt using a preprocessor rule?

https://i.giphy.com/media/WUGNg3FuhiywU/giphy.webp

File: data/news/mediaAssetPage/business-23249447.json [hunk 1]
```diff
@@ -0,0 +1 @@
...:{"short":"A bus driver ruined a perfect shot"},"imageUrl":"ichef.test.bbci.co.uk/images/ic/$recipe/p01ggx88.jpg","embedding":true,"advertising":true,"caption":"A bus driver ruined a perfect shot","versions":[{"versionId":"p01ggx4g","types":["Original"],"duration":38,"durationISO8601":"PT38S","warnings":{},"availableTerritories":{"uk":true,"nonUk":true},"availableFrom":1518788866000}],"type":"media"},"indexImage":{"id":"63722155","subType":"index","href":"http://b.files.bbci.co.uk/D752/test/_63722155_p01ggx88.jpg","path":"/cpsdevpb/D752/test/_63722155_p01ggx88.jpg","height":360,"width":640,"altText":"Keyframe #1","copyrightHolder":"BBC","type":"image"},"id":"urn:bbc:ares::asset:news/av/business-23249447","type":"cps"},"relatedContent":{"section":{"subType":"index","name":"Business","uri":"/news/business","type":"simple"},"site":{"subType":"site","name":"BBC News","uri":"/news","type":"simple"},"groups":[{"type":"more-media-asset-pages","promos":[{"locators":{"assetUri":"/news/10301118","cpsUrn":"urn:bbc:content:assetUri:/news/10301118"},"title":"AV content promotion hyperpuff - importers - LIVE","timestamp":1279018179000,"id":"urn:bbc:ares::asset:news/10301118","type":"hyper"}]}]}}
```
Comment (pjlee11) [inline data/news/mediaAssetPage/business-23249447.json]: Could you run this through a linter please

File: src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import R from 'ramda';
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js]:
```suggestion
import { clone } from 'ramda';
```

File: src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+const buildHeadingBlock = text => ({
+    text,
+    markupType: 'plain_text',
+    type: 'heading',
+  })
+
+export default payload => {
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js]: IMO this is nicer :)
```js
export default payload => {
    const output = clone(payload);

    const headlines = path(['promo', 'headlines'], output);
    const blocks = path(['content', 'blocks'], output);

    if (!headlines || !blocks) return output;

    output.content.blocks = [
        buildHeadingBlock(headingResolver(headlines)),
        ...blocks
    ];

    return output;
};
```

File: src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import R from 'ramda';
+
+export default payload => {
+  const output = R.clone(payload);
+  if (
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js]: Ramda this please :) haha

File: src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import R from 'ramda';
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js]:
```suggestion
import { clone } from 'ramda';
```
I think this helps tree shake too :)

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.test.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    headlines: {
+      headline: 'Default headline',
+    }
+  },
+  content: {
+    blocks: [
+      {
+        text: 'Lonely block :(',
+        markupType: 'plain_text',
+        type: 'text',
```
Comment (pjlee11) [inline src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.test.js]:
```suggestion
        type: 'paragraph',
```

File: data/news/mediaAssetPage/business-23249447.json (hunk 1 above)
Comment (ryanmccombe) [reply data/news/mediaAssetPage/business-23249447.json]: Oops, this usually happens automatically - linted this (and other) files

File: src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  headlines.headline.replace(/<[^>]*>/g, '');
+
+export default payload => {
+  const output = clone(payload);
+
+  const headlines = path(['promo', 'headlines'], output);
+  const blocks = path(['content', 'blocks'], output);
+
+  if (!headlines || !blocks) {
+    return output;
```
Comment (pjlee11) [inline src/app/lib/utilities/preprocessor/rules/addCPSHeadingBlock.js:19]: Do we have tests that cover both passing no `headlines` and no `blocks` to check the `output` is returned?

File: src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import { clone, path } from 'ramda';
+
+export default payload => {
+  const output = clone(payload);
+
+  const blocks = path(['content', 'blocks'], payload);
+  if (!Array.isArray(blocks)) {
+    return output;
```
Comment (pjlee11) [inline src/app/lib/utilities/preprocessor/rules/convertHeadingsToSubheadings.js:8]: Is there a test case covering if `blocks` isn't an array?

**Comment by ryanmccombe** [comment]
@pjlee11 @dr3 is it worth taking the time to split this PR up?  It's quite big and is going to get bigger with these additional tests

**Comment by pjlee11** [comment]
I'm happy with it as is. But you could split it to be `convertHeadingsToSubheadings.js` as one process and then adding the the heading as a seperate process. 

Now that I think about it could we not have `headline` as the main title and then heading doesn't need to change?

**Comment by ryanmccombe** [comment]
That could get messy once we render it, as we will have psammead "subheadings" rendering "headings", and psammead "headings" rendering "headlines"

I think there will be other heading types too?  the data has "crossheadings" 🤷‍♂

**Pull request closed without merging**